src/rt/monitor_.d

```diff
@@ -1,444 +1,289 @@
     }
+    import core.sys.windows.windows;
 
-    extern (C) void _d_monitor_create(Object h)
+    struct Monitor
     {
-        /*
-         * NOTE: Assume this is only called when h.__monitor is null prior to the
-         * call.  However, please note that another thread may call this function
-         * at the same time, so we can not assert this here.  Instead, try and
-         * create a lock, and if one already exists then forget about it.
-         */
-
-        debug(PRINTF) printf("+_d_monitor_create(%p)\n", h);
-        assert(h);
-        Monitor *cs;
-        EnterCriticalSection(&_monitor_critsec);
-        if (!h.__monitor)
-        {
-            cs = cast(Monitor *)calloc(Monitor.sizeof, 1);
-            assert(cs);
-            InitializeCriticalSection(&cs.mon);
-            setMonitor(h, cs);
-            cs.refs = 1;
-            cs = null;
-        }
-        LeaveCriticalSection(&_monitor_critsec);
-        if (cs)
-            free(cs);
-        debug(PRINTF) printf("-_d_monitor_create(%p)\n", h);
+        IMonitor impl; // for user-level monitors
+        DEvent[] devt; // for internal monitors
+        size_t refs; // reference count
+        CRITICAL_SECTION mtx;
     }
 
-    extern (C) void _d_monitor_destroy(Object h)
-    {
-        debug(PRINTF) printf("+_d_monitor_destroy(%p)\n", h);
-        assert(h && h.__monitor && !getMonitor(h).impl);
-        DeleteCriticalSection(&getMonitor(h).mon);
-        free(h.__monitor);
-        setMonitor(h, null);
-        debug(PRINTF) printf("-_d_monitor_destroy(%p)\n", h);
-    }
+    alias Mutex = CRITICAL_SECTION;
+
+    alias initMutex = InitializeCriticalSection;
+    alias destroyMutex = DeleteCriticalSection;
+    alias lockMutex = EnterCriticalSection;
+    alias unlockMutex = LeaveCriticalSection;
+}
+else version (Posix)
+{
+    import core.sys.posix.pthread;
 
-    extern (C) void _d_monitor_lock(Object h)
+    struct Monitor
     {
-        debug(PRINTF) printf("+_d_monitor_acquire(%p)\n", h);
-        assert(h && h.__monitor && !getMonitor(h).impl);
-        EnterCriticalSection(&getMonitor(h).mon);
-        debug(PRINTF) printf("-_d_monitor_acquire(%p)\n", h);
+        IMonitor impl; // for user-level monitors
+        DEvent[] devt; // for internal monitors
+        size_t refs; // reference count
+        pthread_mutex_t mtx;
     }
 
-    extern (C) void _d_monitor_unlock(Object h)
+    alias Mutex = pthread_mutex_t;
+    __gshared pthread_mutexattr_t gattr;
+
+    void initMutex(pthread_mutex_t* mtx)
     {
-        debug(PRINTF) printf("+_d_monitor_release(%p)\n", h);
-        assert(h && h.__monitor && !getMonitor(h).impl);
-        LeaveCriticalSection(&getMonitor(h).mon);
-        debug(PRINTF) printf("-_d_monitor_release(%p)\n", h);
+        pthread_mutex_init(mtx, &gattr);
     }
+
+    alias destroyMutex = pthread_mutex_destroy;
+    alias lockMutex = pthread_mutex_lock;
+    alias unlockMutex = pthread_mutex_unlock;
```

Comment:
These posix functions have non-void return values which are ignored (as in the old version). Is this actually ok?